### ISSUE TITLE
java sym links not updated

### ISSUE THREAD

**Issue by chriscamacho** [voidlinux/void-packages#11006]
### System

* xuname:  
 Void 4.14.14_1 x86_64 GenuineIntel uptodate rrrmF
* package:  
  oracle-jdk 8u162_1

### Expected behavior
java issued from the command line should execute java

### Actual behavior
 instead it says java not found
### Steps to reproduce the behavior
update oracle-jdk
run java

### workaround
xbps-alternatives -g java -s oracle-jdk

additionally JAVA_HOME is not set properly the JAVA_HOME=${JAVA_HOME.... trick doesn't seem to work, removing ${JAVA_HOME and closing } fixes this

**Comment by cr6git** [comment]
> JAVA_HOME is not set properly

This was discussed [here](https://forum.voidlinux.eu/t/java-home-incorrect/290) a long time ago.

**Comment by chriscamacho** [comment]
> This was discussed here a long time ago.
yeah I started the thread, nothing resolved, and JAVA_HOME still not properly set until I fix the profile script every time java updates...

**Comment by cr6git** [comment]
> until I fix the profile script

Please could you describe (precisely) how you fixed it?

**Comment by chriscamacho** [comment]
> removing ${JAVA_HOME and closing } fixes this

so

`JAVA_HOME=${JAVA_HOME......}`

becomes

`JAVA_HOME=......`

please don't let this side issue distract from the main issue of xbps-alternatives symlinks not being updated properly....

**Comment by cr6git** [comment]
I think the real culprit is `openjdk-jre` , see [openjdk/template#L203](https://github.com/voidlinux/void-packages/blob/master/srcpkgs/openjdk/template#L203).

**Comment by chriscamacho** [comment]
i *might* have had the alternatives set to openjdk before the update but can't be sure to be honest

**Comment by cr6git** [comment]
I think we should modify [template#L203](https://github.com/voidlinux/void-packages/blob/master/srcpkgs/openjdk/template#L203) to **force** the creation of a new symlink:
````
ln -sf $_openjdk_version ${PKGDESTDIR}/usr/lib/jvm/openjdk
````

**Comment by chriscamacho** [comment]
is there no way to do

`xbps-alternatives -g java -s <users last choice>`

I have oracle and open versions installed for testing, it would be polite to update the alternatives links to provide the same selection as before the update...

**Comment by chriscamacho** [comment]
just to be clear, it was all the links to each java tool that were broken

/usr/bin/java ,  /usr/bin/javac , /usr/bin/javap etc...

**Comment by cr6git** [comment]
🤔 thinking...

**Comment by newbluemoon** [comment]
In `do_install()` <i>oracle-jdk</i> creates a symlink pointing to the current version

`ln -s jdk${_longVersion} ${DESTDIR}/usr/lib/jvm/oracle-jdk`

But `alternatives` doesn’t make use of this symlink, instead it uses the versioned directory

`java:/usr/bin/java:/usr/lib/jvm/jdk${_longVersion}/bin/java`

I don’t know exactly how the alternatives system works, but it seems to me the symlinks don’t get overwritten/re-created when updating and still point to the old version. I think this wouldn’t be a problem when linking to the non-versioned jdk-symlink like e.g.

`java:/usr/bin/java:/usr/lib/jvm/oracle-jdk/bin/java`

**Comment by chriscamacho** [comment]
I kinda thought that when upgrading previously that it had updated the alternatives links automagically, and if it doesn't then it really oughta !

**Comment by cr6git** [comment]
> In `do_install()` _**oracle-jdk**_ creates a symlink pointing to the current version

ok I see my error... yesterday I mixed up `oracle-jdk` with `openjdk`, I'm sorry.

So, the symlink you talked about is [oracle-jdk/template#L131](https://github.com/voidlinux/void-packages/blob/master/srcpkgs/oracle-jdk/template#L131).
Maybe adding **-f** to force the symlink could be a good idea?

> `java:/usr/bin/java:/usr/lib/jvm/oracle-jdk/bin/java`

About alternatives I think you're right, because they are configured like that in [openjdk/template#L74](https://github.com/voidlinux/void-packages/blob/master/srcpkgs/openjdk/template#L74).

**Comment by cr6git** [comment]
I think it's ok if we modify the INSTALL script [oracle-jdk/INSTALL#L9](https://github.com/voidlinux/void-packages/blob/master/srcpkgs/oracle-jdk/INSTALL#L9), to add this line at the end of `do_install()`:

````
do_install() {
[...]
	# To set the java alternative to 'oracle-jdk'
	xbps-alternatives -s oracle-jdk
}
````
That worked for me.

**Comment by cr6git** [comment]
If I'm correct, my pull request [#11047](https://github.com/voidlinux/void-packages/pull/11047) for `oracle-jre` should be amended accordingly.

**Comment by cr6git** [comment]
> so JAVA_HOME=${JAVA_HOME......} becomes JAVA_HOME=......

Arrff... I forgot about that.

So, you proposed to change [oracle-jdk.sh](https://github.com/voidlinux/void-packages/blob/master/srcpkgs/oracle-jdk/files/oracle-jdk.sh), right?

**Comment by chriscamacho** [comment]
AFAICT

> export JAVA_HOME=${JAVA_HOME=/usr/lib/jvm/oracle-jdk}

wont work, I need to edit each update to:

> export JAVA_HOME=/usr/lib/jvm/oracle-jdk

**Comment by cr6git** [comment]
@chriscamacho ok, I modified [#11047](https://github.com/voidlinux/void-packages/pull/11047) accordingly.

If these changes for `oracle-jre` are correct, now we have to do the same for `oracle-jdk`.

**Comment by cr6git** [comment]
😒 ...looks like I got myself stuck.
